### ISSUE TITLE
Require without Import has side effects on implicits

### ISSUE THREAD

**Issue by PierreCorbineau** [coq/coq#7654]
#### Version 
The Coq Proof Assistant, version 8.8.0 (May 2018)
compiled on May 29 2018 9:57:30 with OCaml 4.06.1

#### Operating system

Linux 4.4.0-127-generic #153-Ubuntu SMP Sat May 19 10:58:46 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux

**Comment by maximedenes** [comment]
I'm having a look at this.

**Comment by maximedenes** [comment]
@ppedrot I think this is a design issue: on one hand, we want to `Require` to have as few effects as possible. On the other hand, why would constants be visible (under qualified names) without their implicit arguments?

**Comment by JasonGross** [comment]
I would like more fine-grained control over side-effect importing.  At the very least, it would be useful for my bug-minimizer if there was a keyword that meant "Require without side-effects", a variant of "Module" that did not export the side-effects that normally follow require, and a keyword that meant "import from this module all the side effects that normally come from require".

**Comment by SkySkimmer** [comment]
If there are no side effects isn't it a noop?

**Comment by JasonGross** [comment]
Without `Require Foo`, `Import Foo` fails.  This is a side-effect I'm okay with `RequireWithoutSideEffects` having.

**Comment by ppedrot** [comment]
@JasonGross It's not hard to implement, but in the current state of the code, expect a lot of anomalies. There are a few key summary states that expect that an available name has been registered, including head and implicit arguments IIRC.

@maximedenes I agree indeed. After discussing with people like @RalfJung who are craving for a proper encapsulation mechanism, I am now essentially convinced that the fact that Require makes available the names of every single CIC object in the corresponding module is a terrible design choice. Not mentionning the fact it is (somewhat) transitive. Problem is, you cannot fix that at this point. At all.

**Comment by ejgallego** [comment]
Pierre-Marie I think we could add a flag that would enable a new implementstion for library, the real issue is that we are missing a huge design task.

**Comment by Blaisorblade** [comment]
Based on the experience with `Hint Resolve`s and the failure of `Set Loose Hint Behavior`, I expect nobody could use the proposed `RequireWithoutSideEffects`.

- It seems like you’d want to add, next to `Global Arguments`, also `Export Arguments`, with similar semantics to `Export Hint Resolve`.
- As a coding guideline, the module defining `foo` should likely stick to `Global Arguments foo` I think. Other modules should stick to `Export Arguments foo` or `Local Arguments foo`. That would address some of the concerns above, including @maximedenes’s, while keeping reasonable semantics:

> On the other hand, why would constants be visible (under qualified names) without their implicit arguments?

(Less reasonable semantics would have `Export Arguments foo` behave globally if `foo` was defined in the same module, whatever that means. IMHO, that’s a clearly inferior option.)

**Comment by Blaisorblade** [comment]
- In fact, I’d want an `Export` version for _all_ commands with `Require`-propagated side effects. I’d even want `Export Set Obligation Tactic`, instead of letting my requires fight.

**Comment by RalfJung** [comment]
`Export Set` already exists, we are using it in Iris.

Is there an exhaustive list of `Require`-propagated side effects? I know of `Hint`, `Instance`, `Set`. What you said above makes it sound like `Arguments` is also in this category?

I think the long-term goal should be for all side-effects to behave the same when using with the same prefix (`Local`/`Global`/`Export` -- maybe `Global` should just be deprecated). Ideally the behavior without a prefix would also be consistent. Being inside a `Section` should not make a difference, because it can be really hard to see (in a long file) if something is in a `Section` or not.

**Comment by ejgallego** [comment]
> Is there an exhaustive list of `Require`-propagated side effects?

No, but it would be easy to actually have a command line flag that would print this information about registered objects, and their scope. Please open an issue if you think that'd be useful.

**Comment by RalfJung** [comment]
I didn't mean on a per-object level, I meant on a per-`Vernacular` level. Or is that the same thing?

This is IMO something that should be part of the docs, it's static information about how Coq works. So I am a bit confused by the proposal to make this be printed dynamically.

**Comment by ejgallego** [comment]
> I didn't mean on a per-object level, I meant on a per-`Vernacular` level. Or is that the same thing?

It is the same thing, some vernaculars define objects that are then stored in vo files and acted upon require / import.

> This is IMO something that should be part of the docs, it's static information about how Coq works. So I am a bit confused by the proposal to make this be printed dynamically.

The right information is already in the ML layer, so such printing could be used to generate the docs automatically, check consistency, etc... Otherwise docs have to done manually.

**Comment by RalfJung** [comment]
I see, I did not realize that these many different side-effects are treated in a unified way by Coq. After all they have some notable inconsistencies in their behavior, which is what lead to my request in the first place.^^

**Comment by RalfJung** [comment]
Opened https://github.com/coq/coq/issues/13414.

**Comment by ejgallego** [comment]
> I see, I did not realize that these many different side-effects are treated in a unified way by Coq. After all they have some notable inconsistencies in their behavior, which is what lead to my request in the first place.^^

There has been recent work to improve that, but indeed we can improve quite a bit more; having the system output what the object scoping rules are is a step towards improvement IMO.

**Comment by Zimmi48** [comment]
First, forget about `Export` and talk about `#[ export ]` instead. It was impossible to transform `Export` into a [legacy attribute](https://coq.inria.fr/refman/language/core/basic.html#grammar-token-legacy-attr) (in the same way that `Local` and `Global` were) because `Export` is also used as a command on its own. Therefore, `Set` / `Unset` will forever remain the only commands that do support the `Export` prefix while the other commands will only support the "modern" attribute syntax (i.e., `#[ export ]`) as is already the case of the `Hint` command family.

Second, yes, it's a mess. Among developers, some commands (with or without the `Global` prefix) have a so-called "superglobal" effect (i.e. the effects happen at `Require` time) and this was the reason to introduce the `export` attribute (which should indeed be extended to more and more commands).

**Comment by Blaisorblade** [comment]
But it seems an improving mess, if the `#[export]` design makes sense for all superglobal effects! But process-wise, it seems it should have been figured that out _before_ starting to advertise the new features — if realistic.
- Can one prove that, without `Global`, `Require B` behaves like `RequireWithoutSideEffects` — so a `Require` in `A` does not affect `A`'s interface, except for `Global` hints? Of course not (`B` is now in scope), but how many counterexamples exist? I'd settle for a published pen-and-paper proof.
- OTOH, 8.12's `#[export] Hint` does sound great, but it'd be unfortunate to migrate to a new design piecewise and find out it doesn't work.

**Comment by SkySkimmer** [comment]
>The right information is already in the ML layer, so such printing could be used to generate the docs automatically, check consistency, etc... 

This is not true, everything is hidden in inversion of control so we have to document manually.

**Comment by ejgallego** [comment]
> This is not true, everything is hidden in inversion of control so we have to document manually.

Not so much anymore, as you have the constructors, which could describe the effects already [and use a tainting mech]

Take my request if you want as also an opportunity to remove some inversion bits.

**Comment by ppedrot** [comment]
> everything is hidden in inversion of control so we have to document manually.

... and libobjects are not even in one-to-one correspondence with vernacular commands, so...

**Comment by ejgallego** [comment]
Indeed we are far from anything resembling an ideal state, the question is if we want to try to improve things, and if so, in which direction. That could be a good project for the CUDW
